x/supplier/keeper/msg_server_stake_supplier.go

```diff
@@ -140,15 +124,46 @@ func (k msgServer) StakeSupplier(ctx context.Context, msg *suppliertypes.MsgStak
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 
-	// Send the coins from the message signer account to the staked supplier pool
 	supplierStakingFee := k.GetParams(ctx).StakingFee
-	stakeWithFee := sdk.NewCoins(coinsToEscrow.Add(*supplierStakingFee))
-	err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, msgSignerAddress, suppliertypes.ModuleName, stakeWithFee)
-	if err != nil {
-		logger.Info(fmt.Sprintf("ERROR: could not send %v coins from %q to %q module account due to %v", coinsToEscrow, msgSignerAddress, suppliertypes.ModuleName, err))
-		return nil, status.Error(codes.InvalidArgument, err.Error())
+
+	// The Supplier is increasing its stake, so escrow the difference
+	if supplierCurrentStake.Amount.LT(msg.Stake.Amount) {
+		coinsToEscrow := msg.Stake.Sub(supplierCurrentStake)
+		stakeWithFee := sdk.NewCoins(coinsToEscrow.Add(*supplierStakingFee))
+
+		// Send the coins from the message signer account to the staked supplier pool
+		err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, msgSignerAddress, suppliertypes.ModuleName, stakeWithFee)
+		if err != nil {
+			logger.Info(fmt.Sprintf("ERROR: could not send %s coins from %q to %q module account due to %v", coinsToEscrow, msgSignerAddress, types.ModuleName, err))
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+
+		logger.Info(fmt.Sprintf("Successfully escrowed %s from %q to %q module account", coinsToEscrow, msgSignerAddress, types.ModuleName))
+	}
+
+	// The supplier is decreasing its stake, unescrow the difference but deduct the staking fee.
+	if supplierCurrentStake.Amount.GT(msg.Stake.Amount) {
+		coinsToUnescrow := supplierCurrentStake.Sub(*msg.Stake)
+
+		if coinsToUnescrow.Amount.LTE(supplierStakingFee.Amount) {
+			err = types.ErrSupplierInvalidStake.Wrapf(
+				"supplier with owner %q must unescrow more than supplier staking fee (%s)",
+				msg.GetOwnerAddress(), supplierStakingFee,
+			)
+			logger.Info(fmt.Sprintf("ERROR: %s", err))
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+
+		coinsAfterFee := sdk.NewCoins(coinsToUnescrow.Sub(*supplierStakingFee))
```

Comment:
Don't we need to add the fee?

Every single time you stake/unstake you pay again. It's like a transaction.

Reply:
The staking fee was incorporated into the amount sent from the supplier's account to the module.

I agree, we should indeed always treat it as an individual transfer in all cases. This way we'll have a distinct transfer event representing the fee.